test/functional/core/main_spec.lua

```diff
@@ -0,0 +1,60 @@
+local lfs = require('lfs')
+local helpers = require('test.functional.helpers')(after_each)
+local global_helpers = require('test.helpers')
+
+local eq = helpers.eq
+local neq = helpers.neq
+local sleep = helpers.sleep
+local nvim_prog = helpers.nvim_prog
+local write_file = helpers.write_file
+
+local popen_w = global_helpers.popen_w
+local popen_r = global_helpers.popen_r
+
+describe('Command-line option', function()
+  describe('-s', function()
+    local fname = 'Xtest-functional-core-main-s'
+    local dollar_fname = '$' .. fname
+    before_each(function()
+      os.remove(fname)
+      os.remove(dollar_fname)
+    end)
+    after_each(function()
+      os.remove(fname)
+      os.remove(dollar_fname)
+    end)
+    it('treats - as stdin', function()
+      eq(nil, lfs.attributes(fname))
+      local pipe = popen_w(
+        nvim_prog, '-u', 'NONE', '-i', 'NONE', '--headless', '-s', '-',
+        fname)
+      pipe:write(':call setline(1, "42")\n')
+      pipe:write(':wqall!\n')
+      pipe:close()
+      local max_sec = 10
+      while max_sec > 0 do
+        local attrs = lfs.attributes(fname)
+        if attrs then
+          eq(#('42\n'), attrs.size)
+          break
+        else
+          max_sec = max_sec - 1
+          sleep(1000)
+        end
+      end
+      neq(0, max_sec)
+    end)
+    it('does not expand $VAR', function()
```

Comment:
This test is there because `-S` is not working well with such names (`-S arg` uses a concat of `source` and `arg` without any escaping used). `-s` is treated different and does not have the problem, but this option looks similar to `-S`.